File: Medium/2401. Longest Nice Subarray.java
class Solution {
    public int longestNiceSubarray(int[] nums) {
        final int N = nums.length;
        int bit = 0, l = 0, res = 0;
        for (int r = 0; r < N; r++) {
            while ((bit & nums[r]) != 0) {
                bit ^= nums[l++];
            }
            bit |= nums[r];
            res = Math.max(res, r - l + 1);
        }
        return res;
    }
}

class Solution1 {
    public int longestNiceSubarray(int[] nums) {
        final int N = nums.length;
        final var bits = new int[32];
        int longest = 1;
        for (int end = 0, start = 0; end < N; end++) {
            update(bits, true, nums[end]);
            while (!check(bits)) {
                update(bits, false, nums[start]);
                start++;
            }
            longest = Math.max(longest, end - start + 1);
        }
        return longest;
    }

    private void update(int[] bits, boolean up, int num) {
        for (int i = 0; i < 32 && num > 0; i++, num /= 2) {
            bits[i] += num % 2 * (up ? 1 : -1);
        }
    }

    private boolean check(int[] bits) {
        for (var bit : bits) {
            if (bit > 1) {
                return false;
            }
        }
        return true;
    }
}

File: Medium/2401. Longest Nice Subarray.py
from typing import List


class BitManipulation:
    def longestNiceSubarray(self, nums: List[int]) -> int:
        bits = 0
        res = 1
        l = 0
        for r, num in enumerate(nums):
            while bits & num:
                bits ^= nums[l]
                l += 1
            res = max(r - l + 1, res)
            bits |= num
        return res


class BruteForce:
    def longestNiceSubarray(self, nums: List[int]) -> int:
        def update(i: int, increment: bool) -> None:
            num = nums[i]
            idx = 0
            while num > 0:
                bits[idx] += num % 2 * (1 if increment else -1)
                num //= 2
                idx += 1
        n, l = len(nums), 0
        res = 1
        bits = [0 for _ in range(32)]
        for r in range(n):
            update(r, True)
            while not all(bit < 2 for bit in bits):
                update(l, False)
                l += 1
            res = max(r - l + 1, res)
        return res